src/erlcloud_ddb1.erl

```diff
@@ -302,6 +313,42 @@ retry(Attempt, Reason) when Attempt >= ?NUM_ATTEMPTS ->
 retry(Attempt, _) ->
     backoff(Attempt),
     {attempt, Attempt + 1}.
-    
+
+-spec are_updates_idempotent(updates()) -> boolean().
+are_updates_idempotent(UpdatesMap) when is_map(UpdatesMap) ->
+    Updates = maps:to_list(UpdatesMap),
+    are_updates_idempotent(Updates);
+are_updates_idempotent([{}]) ->
+    true;
+are_updates_idempotent(Updates) ->
+    lists:all(fun is_update_idempotent/1, Updates).
+
+-spec is_update_idempotent({_, jsx:json_term()}) -> boolean().
+is_update_idempotent({_, Map}) when is_map(Map) ->
+    is_update_content_idempotent(Map);
+is_update_idempotent({_, KvList}) ->
+    Map = maps:from_list(KvList),
+    is_update_content_idempotent(Map).
+
+is_update_content_idempotent(#{<<"Action">> := Action}) ->
+    % https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/API_UpdateItem_v20111205.html#API_UpdateItem_Description
+    case Action of
+        <<"PUT">> -> true;
+        <<"DELETE">> -> true;
+        <<"ADD">> -> false
+    end.
+
+-spec do_opts_keep_idempotency(opts()) -> boolean().
+do_opts_keep_idempotency(OptsMap) when is_map(OptsMap) ->
+    Opts = maps:to_list(OptsMap),
+    do_opts_keep_idempotency(Opts);
+do_opts_keep_idempotency([{}]) ->
+    true;
+do_opts_keep_idempotency(Opts) ->
+    % * https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/API_DeleteItem_v20111205.html#API_DeleteItem_RequestParameters
+    % * https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/API_PutItem_v20111205.html#API_PutItem_RequestParameters
+    % * https://docs.amazonaws.cn/en_us/amazondynamodb/latest/developerguide/API_UpdateItem_v20111205.html#API_UpdateItem_RequestParameters
+    not lists:keymember(<<"Expected">>, 1, Opts).
```

Comment:
This seems wrong. wouldn't setting an expectation (condition) on a request make it more likely to be idempotent?

Reply:
It safeguards you against committing unwanted changes, of course, but it's still a bit weird to get an "expectation failed" sort of error because of a retry - wouldn't it be better to spit the network error back to the consumer, and let them decide? Otherwise it's much more obscure whether to say if the expectation failed because your initial assumptions are wrong (say, because your in-memory cache of the data is outdated) rather than because there was a retry.

Reply:
Fair enough.